### ISSUE TITLE
Grapher's treatment of ADR data is difficult to use

### ISSUE THREAD

**Issue by mgiustina** [labrad/scalabrad-web#320]
When using the Grapher to view ADR data, I've encountered a few issues that cost a lot of time. The following (1, 2, 3) are suggestions for first-order (hopefully quick) improvements; the note at the end describes something a bit more elaborate.

1) The Grapher's default appears to be to load all available dependent data sets on the same plot. This is never desirable for ADR data, which contains temperatures/resistances/voltages/currents with orders of magnitude difference in their ranges.
Suggestion: It would save the user a lot of time if the default traces were always just:
1 - temperature (ch2: 4K) [K]
5 - temperature (ruox) [K]
The user would retain the ability to (de)activate all traces at will.

2) Each time the data-collecting server restarts, a new dataset begins. The grapher plots individual datasets, thus the timespan of a given plot is very arbitrary. If the server suddenly restarted at an inconvenient time, it can be impossible to view today's mag cycle next to yesterday's.
Suggestion: Perhaps a quick fix would be to allow the user to choose a number of previous datasets that would be plotted on the same plot, one after the next (temperature datasets from different days should not be plotted on top of each other). To keep track of which set came from which day, it would be good to indicate where each dataset begins, perhaps with a vertical line intersecting the x-axis.

3) The default x-axis display is in seconds, starting with a value close to 1,500,000,000 (47 years!?). This makes the x-axis very difficult to read and limits the usability of the plots. For certain zoom levels, the labels on neighboring guide lines (e.g. 1,500,000,000 and 1,500,001,000) may appear on top of each other).
Suggestion: Just one idea---the x-axis could continue to display in seconds, but with the beginning of the plot being arbitrarily defined as zero.

Note: It would be nice to easily associate the values on the x-axis with a date on the calendar (and my lab notebook). At the moment, periods of time in which the fridge was open are conveniently omitted from the data. That is nice, because it saves the user from having to wade through periods in which the fridge was warm/open. However it complicates the process of associating a calendar-date to an event in the ADR. A perhaps more ideal solution (perhaps outside the scope of the Grapher, idk) would be:
- To use the grapher, the user commands the plotting to cover a certain date range
- Associated with the beginning of each cooldown in that date range, there is an absolute timestamp.
- The x-axis is delineated in seconds. Each cooldown has its own x=0 seconds at the beginning of the cooldown, which is the absolute timestamp (from the beginning of each cooldown). So the x-axis has multiple points called "0", one per cooldown.

(@maffoo, @DanielSank, please let me know if this is clear or if a picture is needed for clarity.)

**Comment by DanielSank** [comment]
@joshmutus Is `#3` something you could implement easily?

**Comment by DanielSank** [comment]
I think `#1` could be addressed in one of two ways:

1. Put a `which traces to plot` parameter in the data set. The grapher would show those traces when the data set is loaded. We've talked about supporting plotting hints before. I think this is a good idea.

1. Store plot hints on the client.

@maffoo ?

**Comment by joshm-3b6b58** [comment]
@DanielSank for #3, we should just change the ADR server, that's much easier.

**Comment by DanielSank** [comment]
How can we solve this problem via the ADR server?

Also, it's a bit concerning that changing the grapher is "hard". Is there an easy to explain reason?

**Comment by joshm-3b6b58** [comment]
By subtracting the initial time from all the time values it saves.

**Comment by joshm-3b6b58** [comment]
@DanielSank I never said it was hard. 

Maybe I'm not understanding the proposed solution, but it just seems arbitrary to add a rule about how one kind of data is displayed. If we want the data to displayed differently then we should save it differently.

**Comment by DanielSank** [comment]
I strongly disagree that we should handle plotting by changing our raw data format.

**Comment by joshm-3b6b58** [comment]
I'm not saying change the format, I'm just saying change the zero. The time since Jan 1 1970 is irrelevant for this plot, the time since the start of the cooldown is what we care about, we should save that.

**Comment by DanielSank** [comment]
So, what do we change the zero to?

**Comment by joshm-3b6b58** [comment]
The beginning of the cooldown when the ADR server starts logging temperatures.

**Comment by joshm-3b6b58** [comment]
We could certainly add conversion of UTC time to date time as an option for plotting, it just would be awkward until we build in some plot helpers saved with the dataset.

**Comment by DanielSank** [comment]
> The beginning of the cooldown when the ADR server starts logging temperatures.

But then when I look at the time axis and see "34 hours", to make sense of it I have to figure out when the cooldown started and do arithmetic.

> We could certainly add conversion of UTC time to date time as an option for plotting

Yes!

**Comment by joshm-3b6b58** [comment]
> But then when I look at the time axis and see "34 hours", to make sense of it I have to figure out when the cooldown started and do arithmetic.

Maybe I'm missing something, but I thought that was exactly what we wanted to see.

The problem we have with adding UTC time and date is that how will the grapher know when to plot it. Very little our data is stored as UTC times. Do we put a radio button? It seems silly to me to put a radio button that's applicable for something like 0.01% of our datasets. If we plumb in plot helpers this would work great. I can start the plumbing from the grapher side so it has the ability to convert to a date time, but the real question is when to apply that.

**Comment by DanielSank** [comment]
>> But then when I look at the time axis and see "34 hours", to make sense of it I have to figure out when the cooldown started and do arithmetic.

> Maybe I'm missing something, but I thought that was exactly what we wanted to see.

Perhaps. Then maybe for another use case I want to know the actual date/time to compare with my lab notebook. So now you're saying I have to save a copy of time axis in a special format for every possible way I'd want to look at the times, i.e. `time_referred_to_cooldown_start` and `time_as_datetime`. That's obviously not a good idea. We want one set of data, probably in UTC, and the ability to *view* it in as many ways as is useful.

**Comment by joshm-3b6b58** [comment]
Right but the ability to view data in many useful ways is a feature that's larger in scope than this simple cooldown dataset, and to be properly handled requires some additional data to be saved with the dataset regardless. For example, how would the grapher know a time (in seconds) is a UTC timecode? What if the user decided to save they're data with 0 being the beginning of a scan?

**Comment by DanielSank** [comment]
> Right but the ability to view data in many useful ways is a feature that's larger in scope than this simple cooldown dataset

I dunno, supporting a few time display formats seems generally useful *and* the right solution here.

> and to be properly handled requires some additional data to be saved with the dataset regardless. For example, how would the grapher know a time (in seconds) is a UTC timecode? What if the user decided to save they're data with 0 being the beginning of a scan?

If I saved my data with 0 being the beginning of the scan, then I simply don't use the UTC->datetime view option.

**Comment by joshm-3b6b58** [comment]
> I dunno, supporting a few time display formats seems generally useful and the right solution here.

I agree supporting them is useful, I'm debating on _how_ we support them. If we want to store extra metadata with the datasets that help with plotting (default traces, etc) then that's how I think this should be done. I'm worried that there's a ton of view options people will want to tack on and then we'll have trouble handling them all in the UI

**Comment by DanielSank** [comment]
Default traces and selectable time formats seem like completely different features to me.

Default traces must be specified inside the data set. However, I might want to look at the time axis in more than one way, and I need to be able to switch back and forth, so that has to be supported by the UI.

I don't see adding a few time display formats as bloat. It's a totally reasonable thing you want to do with time data.

**Comment by joshm-3b6b58** [comment]
Yeah, guess I'm coming around, can you file a separate issue and we can discuss which formatting variants are useful for the x-axis?

**Comment by DanielSank** [comment]
Done: https://github.com/labrad/scalabrad-web/issues/321
